File: C/libsrc/ktpnik.c
#include <ktpnik.h>

#define PRIVATE __attribute__((visibility("hidden")))

// MINIMUM DATA = Sumatera, Aceh, Kab Aceh Selatan, Kec Bakogan, Laki-Laki, 01/01/2000, Index 0001
const uint64_t const 
PRIVATE MIN_NIK_INFO =     1101010101000001ULL,
PRIVATE SMP_NIK_INFO =     3502136410016213ULL,
PRIVATE NIK_PROVINCE_DIV =  100000000000000ULL,
PRIVATE NIK_CITY_DIV =        1000000000000ULL,
PRIVATE NIK_REGION_DIV =        10000000000ULL,
PRIVATE NIK_DATE_DIV =            100000000ULL,
PRIVATE NIK_MONTH_DIV =             1000000ULL,
PRIVATE NIK_YEAR_DIV =                10000ULL,
PRIVATE NIK_REGISTNUM_DIV =               1ULL,
PRIVATE NIK_HUNDRED_MAX =                100ULL,
PRIVATE NIK_REGISTNUM_MAX =            10000ULL,
PRIVATE NIK_ISFEMALE_FLAG =              40ULL
;

#define __min(a, b) (((a) < (b)) ? (a) : (b))
#define __max(a, b) (((a) > (b)) ? (a) : (b))
#define __clm(a, b) ((a) % (b))

/**
 * Parse NIK Number to [KtpNik]
 * @param nik - Source KTP NIK
 * @param retval - Reference to a NIK Structure
 * @returns Error type, [KTPNIK_SUCCESS] if success
 */
enum ktpnik_result ktpnik_parse(uint64_t nik, struct ktpnik_t* ktp_nik){
	if(nik < MIN_NIK_INFO) return KTPNIK_ERROR_INSUFFICIENT_DATA;
	struct ktpnik_t* k = ktp_nik;
	
	k->location.province = (uint8_t)__clm(nik / NIK_PROVINCE_DIV, NIK_HUNDRED_MAX);
	k->location.city = (uint8_t)__clm(nik / NIK_CITY_DIV, NIK_HUNDRED_MAX);
	k->location.region = (uint8_t)__clm(nik / NIK_REGION_DIV, NIK_HUNDRED_MAX); 
	
	k->birthday.date = (uint8_t)__clm(nik / NIK_DATE_DIV, NIK_HUNDRED_MAX);
	k->birthday.month = (uint8_t)__clm(nik / NIK_MONTH_DIV, NIK_HUNDRED_MAX);
	k->birthday.year = (uint8_t)__clm(nik / NIK_YEAR_DIV, NIK_HUNDRED_MAX);
	
	if((k->is_female = k->birthday.date > NIK_ISFEMALE_FLAG) == 1){
		k->birthday.date -= NIK_ISFEMALE_FLAG;
	}
	
	k->person_unique_id = (uint16_t)__clm(nik / NIK_REGISTNUM_DIV, NIK_REGISTNUM_MAX);
	
	if(k->birthday.date > 31 || k->birthday.month > 12) return KTPNIK_ERROR_INVALID_DATE;
	
	return KTPNIK_SUCCESS;
}

File: Java/ktpnik/main/src/emiyasyahriel/ktpnik/KtpNik.java
package emiyasyahriel.ktpnik;
/**
 * ktpnik - KTP NIK Parsing Library
 * Author : EmiyaSyahriel
 */

import java.math.BigInteger;

/**
 * Data Structure of a KTP NIK
 */
public class KtpNik {
    
    /** Register Province */ 
    public byte province;
    /** Register City / Regency */
    public byte city;
    /** Register Region */
    public byte region;
    /** Is Resident Female? */
    public Boolean isFemale;
    /** Birth Date */
    public byte birthDate;
    /** Birth Month */
    public byte birthMonth;
    /** Birth Year */
    public byte birthYear;
    /** Person Unique ID */
    public short uid;

    private static BigInteger MIN_NIK_INFO = new BigInteger("1101010101000001");

    private static BigInteger NIK_PROVINCE_DIV =  new BigInteger("100000000000000");
    private static BigInteger NIK_CITY_DIV =      new BigInteger("1000000000000");
    private static BigInteger NIK_REGION_DIV =    new BigInteger("10000000000");
    private static BigInteger NIK_DATE_DIV =      new BigInteger("100000000");
    private static BigInteger NIK_MONTH_DIV =     new BigInteger("1000000");
    private static BigInteger NIK_YEAR_DIV =      new BigInteger("10000");
    private static BigInteger NIK_REGISTNUM_DIV = new BigInteger("1");
    private static BigInteger NIK_HUNDRED_MAX =   new BigInteger("100");
    private static BigInteger NIK_REGISTNUM_MAX = new BigInteger("10000");

    /**
     * Parse NIK Number to [KtpNik]
     * @param nik - Source KTP NIK
     * @param retval - Reference to a NIK Structure
     * @returns Error type, {KtpNikError} if success
     */
    public static KtpNikError Parse(BigInteger nik, KtpNik retval){
        if (nik.compareTo(MIN_NIK_INFO) < 0)
        {
            return KtpNikError.InsufficientData;
        }        
        
        retval.province =   nik.divide(NIK_PROVINCE_DIV).mod(NIK_HUNDRED_MAX).byteValue();
        retval.city =       nik.divide(NIK_CITY_DIV).mod(NIK_HUNDRED_MAX).byteValue();
        retval.region =     nik.divide(NIK_REGION_DIV).mod(NIK_HUNDRED_MAX).byteValue();
        retval.birthDate =  nik.divide(NIK_DATE_DIV).mod(NIK_HUNDRED_MAX).byteValue();
        retval.birthMonth = nik.divide(NIK_MONTH_DIV).mod(NIK_HUNDRED_MAX).byteValue();
        retval.birthYear =  nik.divide(NIK_YEAR_DIV).mod(NIK_HUNDRED_MAX).byteValue();
        retval.uid =        nik.divide(NIK_REGISTNUM_DIV).mod(NIK_REGISTNUM_MAX).shortValue();
        retval.isFemale = retval.birthDate > 40;

        if (retval.isFemale)
        {
            retval.birthDate -= 40;
        }

        if (retval.birthDate > 31 || retval.birthMonth > 12)
        {
            return KtpNikError.InvalidDate;
        }
        
        return KtpNikError.Success;
    }
}